src/bridge/events.rs

```diff
@@ -159,494 +160,458 @@ fn unpack_color(packed_color: u64) -> Color4f {
     }
 }
 
-fn parse_array(array_value: &Value) -> Result<&[Value]> {
-    if let Value::Array(content) = array_value {
-        Ok(&content[..])
-    } else {
-        Err(EventParseError::InvalidArray(array_value.clone()))
-    }
+fn pop_value(values: &mut Vec<Value>) -> Result<Value> {
+    values.pop().ok_or(EventParseError::InvalidEventFormat)
 }
 
-fn parse_map(map_value: &Value) -> Result<&[(Value, Value)]> {
-    if let Value::Map(content) = map_value {
-        Ok(&content[..])
-    } else {
-        Err(EventParseError::InvalidMap(map_value.clone()))
+macro_rules! extract_values {
+    ([$($i: ident),+], $values: expr $(,)?) => {
```

Comment:
I feel pretty strongly that we should only use macros when there is no obvious way to implement the functionality without them or if the readability of the code goes significantly up.

In this case there is a clear and obvious readability gain, but I'm not sure that we couldn't have solved this with a simple function. For example, what if instead we wrote a function which took a value and an expected number of arguments and returned a result which is the array of that number of arguments if the count matches, or the appropriate error otherwise.

Such an function could be used like this:

``` Rust
let [_cursor_style_enabled, mode_info] = extract_values(mode_info_set_arguments, 2)?;
```

Then when someone comes along to read the code they fully understand the contract and when jumping to the definition, its familiar code instead of the relatively hard to parse macro syntax. Does that make sense?

Reply:
Humm, maybe I can do something as you suggested. It wouldn't be exactly as you suggested, as the size of an array should be known at compile-time, but I think we should be able to write a function that does this.

Reply:
Maybe a dumb question, but why does the size of the array need to be known at compile time? Oh. So the de-structuring works?

Reply:
> why does the size of the array need to be known at compile time?

Because the compiler needs to know the size of the array to know how much memory must the reserved for it on the stack.